xblock/scorable.py

```diff
@@ -0,0 +1,143 @@
+"""
+Scorable.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from collections import namedtuple
+import six
+
+
+Score = namedtuple('Score', ['earned', 'total'])
+
+
+class ScorableXBlockMixin(object):
+    """
+    Mixin to handle functionality related to scoring.
+
+    Provides the following public methods:
+
+        ScorableXBlockMixin.rescore(self, only_if_higher)
+            Calculate a new score and save it to the block.
+            `only_if_higher
+
+        ScorableXBlockMixin.allows_rescore(self) -> bool
+            Returns True if the block can be rescored.
+
+    Subclasses must define the following:
+
+        ScorableXBlockMixin.get_score(self)
+            Return the currently saved score.  Must not calculate a new score.
+
+        ScorableXBlockMixin.set_score(self, score)
+            Set the score on the block to the specified value
+
+        ScorableXBlockMixin.calculate_score(self, score)
+            Calculate a new score based on the state of the block.  Must not
+            update any values on the block.
+    """
+
+    def rescore(self, only_if_higher):
+        """
+        Calculate a new score and save it to the block.  If only_if_higher is
+        True and the score didn't improve, keep the existing score.
+
+        Returns True if the score was changed.
+        Returns False if the score was not changed.
+        Raises a TypeError if the block cannot be scored.
+        Raises a ValueError if the user has not yet completed the problem.
+        """
+
+        _ = self.runtime.service(self, 'i18n').ugettext
+        if not self.allows_rescore():
+            self._publish_scoring_event('rescore_failure', {'failure': 'unsupported'})
```

Comment:
Agreed.  I was slow on the uptake between event publishing and logging, because Capa publishes events for everything, but I'm not sure we need to do that.